src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Resolvers/CompositeResolver.cs

```diff
@@ -38,124 +30,102 @@ public CompositeResolver()
+                        if (!formattersByType.ContainsKey(ti.GenericTypeArguments[0]))
+                        {
+                            formattersByType.Add(ti.GenericTypeArguments[0], formatter);
+                        }
                     }
                 }
-            }
 
-            if (!foundAny)
-            {
-                throw new ArgumentException("No formatters found on this object.", nameof(formatter));
+                if (!foundAny)
+                {
+                    throw new ArgumentException("No formatters found on this object: " + formatter.GetType().FullName, nameof(formatter));
+                }
             }
+
+            // Make a copy of the resolvers list provided by the caller to guard against them changing it later.
+            return new CachingResolver(
+                new ReadOnlyDictionary<Type, object>(formattersByType),
+                new ReadOnlyCollection<IFormatterResolver>(resolvers.ToArray()));
         }
 
-        /// <inheritdoc/>
-        public IMessagePackFormatter<T> GetFormatter<T>()
+        private class CachingResolver : CachingFormatterResolver
         {
-            // We use locks here because this resolver can be called by multiple threads concurrently
-            // and we may be mutating some of these collections as we cache formatters per type.
-            object formatter;
-            lock (this.formattersByType)
-            {
-                if (this.formattersByType.TryGetValue(typeof(T), out formatter))
-                {
-                    return (IMessagePackFormatter<T>)formatter;
-                }
-            }
+            private readonly ReadOnlyDictionary<Type, object> formattersByType;
+            private readonly ReadOnlyCollection<IFormatterResolver> subResolvers;
 
-            foreach (IFormatterResolver resolver in this.subResolvers)
+            /// <summary>
+            /// Initializes a new instance of the <see cref="CachingResolver"/> class.
+            /// </summary>
+            internal CachingResolver(ReadOnlyDictionary<Type, object> formattersByType, ReadOnlyCollection<IFormatterResolver> subResolvers)
             {
-                formatter = resolver.GetFormatter<T>();
-                if (formatter != null)
-                {
-                    break;
-                }
+                this.formattersByType = formattersByType ?? throw new ArgumentNullException(nameof(formattersByType));
+                this.subResolvers = subResolvers ?? throw new ArgumentNullException(nameof(subResolvers));
             }
 
-            // Remember the answer for next time.
-            lock (this.formattersByType)
+            /// <inheritdoc/>
+            protected override IMessagePackFormatter<T> GetFormatterCore<T>()
             {
-                if (!this.formattersByType.ContainsKey(typeof(T)))
+                if (!this.formattersByType.TryGetValue(typeof(T), out object formatter))
```

Comment:
most cases only add Resolver so create `CachingResolver` per overload to remove unnecessary dictionary lookup.

Reply:
Done. Although I doubt an *empty* dictionary lookup costs anything significant.